gen/calc.py

```diff
@@ -519,6 +529,44 @@ def validate_exhibitor_storage_master_discovery(master_discovery, exhibitor_stor
             "`master_http_load_balancer` then exhibitor_storage_backend must not be static."
 
 
+def validate_dns_bind_ip_blacklist(dns_bind_ip_blacklist):
+    return validate_ip_list(dns_bind_ip_blacklist)
+
+
+def validate_dns_forward_zones(dns_forward_zones):
+    """
+     "forward_zones": [["a.contoso.com", [["1.1.1.1", 53],
+                                          ["2.2.2.2", 53]]],
+                       ["b.contoso.com", [["3.3.3.3", 53],
+                                          ["4.4.4.4", 53]]]]
+    """
+    zone_defs = None
+    try:
+        zone_defs = json.loads(dns_forward_zones)
+    except ValueError as ex:
+        raise AssertionError(fz_err("{} is not valid JSON: {}".format(dns_forward_zones, ex))) from ex
+    assert isinstance(zone_defs, list), fz_err("{} is not a list".format(zone_defs))
+
+    for z in zone_defs:
+        assert isinstance(z, list), fz_err("{} is not a list".format(z))
+        assert len(z) == 2, fz_err("{} is not length 2".format(z))
+        assert isinstance(z[0], str), fz_err("{} is not a string".format(z))
+
+        upstreams = z[1]
+        for u in upstreams:
+            assert isinstance(u, list), fz_err("{} not a list".format(u))
+            assert len(u) == 2, fz_err("{} not length 2".format(u))
+
+            ip = u[0]
+            port = u[1]
+            assert valid_ipv4_address(ip), fz_err("{} not a valid IP address".format(ip))
+            validate_int_in_range(port, 1, 65535)
+
+
+def fz_err(msg):
```

Comment:
It'd be nice to define this within the body of `validate_dns_forward_zones()`, since it's only relevant there.